cluster/tls_transport_test.go

```diff
@@ -0,0 +1,213 @@
+// Copyright 2020 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cluster
+
+import (
+	"bufio"
+	context2 "context"
+	"fmt"
+	"io"
+	"net"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/stretchr/testify/require"
+)
+
+var logger = log.NewNopLogger()
+
+func TestNewTLSTransport(t *testing.T) {
+	testCases := []struct {
+		bindAddr    string
+		bindPort    int
+		tlsConfFile string
+		err         string
+	}{
+		{err: "must specify TLSTransportConfig"},
+		{err: "invalid bind address \"\"", tlsConfFile: "testdata/tls_config_node1.yml"},
+		{bindAddr: "abc123", err: "invalid bind address \"abc123\"", tlsConfFile: "testdata/tls_config_node1.yml"},
+		{bindAddr: localhost, bindPort: 0, tlsConfFile: "testdata/tls_config_node1.yml"},
+		{bindAddr: localhost, bindPort: 9094, tlsConfFile: "testdata/tls_config_node2.yml"},
+	}
+	l := log.NewNopLogger()
+	for _, tc := range testCases {
+		cfg := mustTLSTransportConfig(tc.tlsConfFile)
+		transport, err := NewTLSTransport(context2.Background(), l, nil, tc.bindAddr, tc.bindPort, cfg)
+		if len(tc.err) > 0 {
+			require.Equal(t, tc.err, err.Error())
+			require.Nil(t, transport)
+		} else {
+			require.Nil(t, err)
+			require.Equal(t, tc.bindAddr, transport.bindAddr)
+			require.Equal(t, tc.bindPort, transport.bindPort)
+			require.Equal(t, l, transport.logger)
+			require.NotNil(t, transport.listener)
+			transport.Shutdown()
+		}
+	}
+}
+
+const localhost = "127.0.0.1"
+
+func TestFinalAdvertiseAddr(t *testing.T) {
+	testCases := []struct {
+		bindAddr      string
+		bindPort      int
+		inputIp       string
+		inputPort     int
+		expectedIp    string
+		expectedPort  int
+		expectedError string
+	}{
+		{bindAddr: localhost, bindPort: 9094, inputIp: "10.0.0.5", inputPort: 54231, expectedIp: "10.0.0.5", expectedPort: 54231},
+		{bindAddr: localhost, bindPort: 9093, inputIp: "invalid", inputPort: 54231, expectedError: "failed to parse advertise address \"invalid\""},
+		{bindAddr: "0.0.0.0", bindPort: 0, inputIp: "", inputPort: 0, expectedIp: "random"},
```

Comment:
Such a test is not very packager-friendly. Most distros will run unit tests in clean room environment when packaging, which can often mean a host with only a loopback interface (with 127.0.0.1 and _usually_ [::1] configured), and no default route.

On Debian buildbots, and presumably other distros with similar clean room package building policy, this fails with:
```
=== RUN   TestFinalAdvertiseAddr
    tls_transport_test.go:89: 
        	Error Trace:	/<<PKGBUILDDIR>>/build/src/github.com/prometheus/alertmanager/cluster/tls_transport_test.go:89
        	Error:      	Expected nil, but got: &errors.errorString{s:"no private IP address found, and explicit IP not provided"}
        	Test:       	TestFinalAdvertiseAddr
--- FAIL: TestFinalAdvertiseAddr (0.00s)
```
This is due to the `bindAddr` being "0.0.0.0" and the `inputIp` being empty.

I see that TestClusterJoinAndReconnect in cluster/cluster_test.go at least checks first whether a private IP exists, and skips the test if none is found. Perhaps it makes sense to omit this particular test case in such a scenario also.

Reply:
> I see that TestClusterJoinAndReconnect in cluster/cluster_test.go at least checks first whether a private IP exists, and skips the test if none is found. Perhaps it makes sense to omit this particular test case in such a scenario also.

I agree and the skip in cluster_test.go exists exactly for the same constraint (https://github.com/prometheus/alertmanager/issues/1445).